### ISSUE TITLE
DEP: extend deprecation of private namespaces also to fortran-generated modules

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#19907]
While looking at cleaning up the deprecations following #18279 (fitting into the larger picture of #14360), I noticed that `scipy.interpolate.dfitpack` is not deprecated, which seems like an oversight - in particular because it contains several functions that `scipy.interpolate` itself contains:
```python
>>> import scipy.interpolate
>>> import scipy.interpolate.dfitpack
>>> dfitpack_funcs = set(scipy.interpolate.dfitpack.__dir__())
>>> interpolate_funcs = set(scipy.interpolate.__dir__())
>>> sorted([x for x in dfitpack_funcs & interpolate_funcs if not x.startswith("_")])
['spalde', 'splder', 'splev', 'splint', 'sproot']
```

But that's only half the story, because we have a new interface to fitpack...
https://github.com/scipy/scipy/blob/52732578a67e1f243b466e0668d99f6ec85d0d08/scipy/interpolate/__init__.py#L170-L171
and funnily enough that [file](https://github.com/scipy/scipy/blob/main/scipy/interpolate/_fitpack2.py) wraps _both_ `dfitpack` directly, as well as a separate wrapper [layer](https://github.com/scipy/scipy/blob/main/scipy/interpolate/_fitpack_impl.py), also around `dfitpack` (sidenote: looks like this [tweet](https://twitter.com/sc_codeUM/status/1747684397745049964) I saw today turned out to be prophetic 😅 ). Suffice it to say, despite different names, a lot (all?) of the remaining functions in `dfitpack` then also get accounted for.

It's looks likely that this was overlooked because there's no actual python code, we just directly generate the functions from fortran:
https://github.com/scipy/scipy/blob/52732578a67e1f243b466e0668d99f6ec85d0d08/scipy/interpolate/meson.build#L145-L162

Not sure how many other such should-be-private-but-has-no-underscore fortran libraries are about.

CC @rgommers @tupui @j-bowhay @ev-br

**Comment by ev-br** [comment]
The story here is a bit different.

TL;DR: 1) `scipy.interpolate.dfitpack` is an implementation detail, not meant to be public; 2) any "new" comment is likely obsolete, can be removed if wanted; 3) cleanup is ongoing, help appreciated, the work is way beyond just deprecating accidentally public names.

Long version. Let's start from the source files:

```
$ ll scipy/interpolate
<edited for clarity>

drwxrwxr-x  2 br br  4096 янв 17 15:20 fitpack/    # FORTRAN sources
-rw-rw-r--  1 br br 89172 янв  7 15:24 _fitpack2.py    # implementation details of `UnivariateSpline` etc
-rw-rw-r--  1 br br   964 ноя 22 19:48 fitpack2.py    # deprecated wrappers for _fitpack2.py
-rw-rw-r--  1 br br 28669 янв  7 15:24 _fitpack_impl.py    # implementation details of `splrep` etc
-rw-rw-r--  1 br br   716 ноя 22 19:48 fitpack.py    # deprecated wrappers for fitpack.py
-rw-rw-r--  1 br br 27528 янв 17 15:20 _fitpack_py.py    # implementation details for `splrep` etc, goes with `_fitpack_impl.py`

(scipy-dev) br@gonzales:~/repos/scipy/scipy$ ll scipy/interpolate/src/
-rw-rw-r-- 1 br br  1696 ноя 21  2022 __fitpack.h               # is unrelated to FORTRAN
-rw-rw-r-- 1 br br 19315 янв  6 13:11 _fitpackmodule.c          # sources for the `_fitpack.so` extension
-rw-rw-r-- 1 br br 32669 ноя 22 19:48 fitpack.pyf           # sources for the `dfitpack.so` extension
```

The fortran sources in `scipy/interpolate/fitpack` have two sets of C wrapper to python. The extension modules are `_fitpack.so` and `dfitpack.so`. 

The comment "New interface to fitpack library" is historical (remember book titles like "modern analysis of ..."?) is from 2003:

```
$ git blame -M -C -C scipy/interpolate/__init__.py
...
2715c638314 Lib/interpolate/__init__.py   (Pearu Peterson  2003-09-15 13:58:12 +0000 171) # New interface to fitpack library:
23ef8d5c336 scipy/interpolate/__init__.py (Gagandeep Singh 2021-10-27 15:42:25 +0530 172) from ._fitpack2 import *
...
```

Why there are two extensions: historical reasons, I suppose. 
Atsushi Sakai did a lot of work deduplicating them recently, see https://github.com/scipy/scipy/issues/16729.

Neither of these extensions needs to be public, anything they export is an implementation detail of either `splrep` et al or `UnivariateSpline` and friends. 
Some dfitpack functions share the name but differ from eponymous public functions scipy.interpolate: e.g. `interpolate.splder` does not use `dfitpack.splder`.

Whether it is worth the hassle to make `dfitpack` raise a DeprecationWarning, I don't know. I personally am -0 (== am not going to do it myself; do not believe it is worth doing;  will not oppose if somebody think it's worth and does it).

A more productive line of work is IMO to cook up a feature-complete alternative to FITPACK with an eye to remove the latter at some point. 
The exact path to removal needs to be worked out. 
Initial steps are tracked in https://github.com/users/ev-br/projects/1/views/1, the currently open scipy PRs are 
https://github.com/scipy/scipy/pull/19753, https://github.com/scipy/scipy/pull/19777 and https://github.com/scipy/scipy/pull/19873
Help is welcome and appreciated :-).

**Comment by ev-br** [comment]
Looks like dfitpack -> _dfitpack rename was done in https://github.com/scipy/scipy/pull/20612, is there something else for this issue @h-vetinari ?

**Comment by ev-br** [comment]
The last one is/was `interpolate/interpnd.so`, so here's gh-21754 to close the book.
